<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="dcterms.date" content="2024-12-09">

<title>Mapping the Proteome-Phenome Atlas: A New Frontier in Precision Medicine</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../styles.css">
</head>

<body>

<div id="quarto-search-results"></div>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Mapping the Proteome-Phenome Atlas: A New Frontier in Precision Medicine</h1>
  <div class="quarto-categories">
    <div class="quarto-category">Interviews</div>
  </div>
  </div>



<div class="quarto-title-meta">

    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">December 9, 2024</p>
    </div>
  </div>
  
    
  </div>
  

</header>




    <link href="https://fonts.googleapis.com/css2?family=Roboto:wght@400;700&amp;display=swap" rel="stylesheet">

    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Mapping the Proteome-Phenome Atlas: A New Frontier in Precision Medicine</title>

    <!-- Open Graph / Facebook -->
    <meta property="og:type" content="article">
    <meta property="og:url" content="https://statsupai.com/pages/S11.html">
    <meta property="og:title" content="Mapping the Proteome-Phenome Atlas: A New Frontier in Precision Medicine">
    <meta property="og:description" content="">
    <meta property="og:image" content="figures/S11_1.jpg">
    <style>
        body {
            font-family: 'Roboto', sans-serif;
            margin: 0;
            padding: 0;
            background-color: #f4f4f4;
            line-height: 1.6;
        }
        .interviewee-introduction {
            display: flex;
            align-items: center;
            background-color: #f8f8f8; /* 轻灰色背景 */
            border-radius: 8px; /* 边角圆润 */
            padding: 20px;
            margin-bottom: 20px; /* 和问答部分保持一定距离 */
        }

        .interviewee-photo {
            width: 150px; /* 根据需要调整 */
            height: 150px; /* 根据需要调整 */
            border-radius: 50%; /* 创建圆形图片 */
            object-fit: cover; /* 确保图片内容适应尺寸 */
            margin-right: 20px;
        }

        .interviewee-details h2 {
            margin-top: 0;
            color: #082653; /* 深蓝色，你可以根据网站主题更改 */
        }

        .interviewee-details p {
            color: #333; /* 深灰色文本 */
            line-height: 1.6; /* 增加行距 */
        }

        .header-image {
            width: 100%; /* 设置宽度为文本宽度的 80% */
            max-width: 1400px; /* 设置图片的最大宽度，与文本区域匹配 */
            height: auto; /* 确保图片比例不失真 */
            display: block; /* 使图片成为块级元素 */
            margin: 20px auto; /* 居中对齐，增加上下间距 */
            border-radius: 8px; /* 圆角处理，增加柔和感 */
            box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1); /* 轻微阴影，减轻视觉重量 */
            border: 3px solid #ddd; /* 边框颜色与页面风格统一 */
        }
        .inline-image {
            width: 85%; /* 设置宽度为容器宽度的一半，可根据需要调整 */
            max-width: 600px; /* 限制图片最大宽度 */
            height: auto; /* 保持图片比例 */
            display: block; /* 将图片作为块级元素显示 */
            margin: 10px auto; /* 居中对齐，增加间距 */
            border-radius: 4px; /* 圆角处理 */
            box-shadow: 0 2px 4px rgba(0, 0, 0, 0.1); /* 添加轻微阴影 */
            border: 2px solid #ccc; /* 边框颜色柔和，与文字风格一致 */
        }

        .interview {
            margin: 10px;
            background-color: rgba(255, 255, 255, 0.9);
            padding: 20px;
            border-radius: 8px;
            box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1);
        }

        .answer-section {
            margin-bottom: 30px;
            padding-bottom: 20px;
            border-bottom: 2px solid #ccc;
        }

        .question-title {
            font-size: 1.2rem;
            color: #003366;
            font-weight: bold;
            margin-bottom: 15px;
        }

        .answer-paragraph {
            margin-bottom: 15px;
            background-color: #f9f9f9;
            padding: 10px;
            border-radius: 5px;
            line-height: 1.8;
            box-shadow: 0 2px 4px rgba(0, 0, 0, 0.1);
        }

        .answer-paragraph:last-child {
            margin-bottom: 0;
        }

        .edit-info {
            color: #808080;
            font-size: 15px;
            text-align: right;
            padding: 20px 0;
            margin-top: 20px;
            border-top: 1px solid #ccc;
        }

        .article-link {
            background-color: #f4f4f4;
            padding: 10px;
            border-radius: 8px;
            text-align: center;
            margin: 20px 0;
        }

        .article-link p {
            margin: 0;
            color: #003366;
            font-weight: bold;
        }

        .article-link a {
            display: inline-block;
            background-color: #003366;
            color: white;
            padding: 10px 20px;
            border-radius: 5px;
            text-decoration: none;
            font-size: 18px;
            margin-top: 10px;
            transition: background-color 0.3s;
        }

        .article-link a:hover {
            background-color: #0056b3;
        }
        
        .text-image-container {
            display: flex;
            align-items: flex-start; /* 对齐方式 */
            gap: 20px; /* 设置文字和图片之间的间距 */
            margin: 20px 0; /* 增加上下间距 */
        }

        .text-content {
            flex: 1; /* 占用容器的一部分，自动调整宽度 */
        }

        .image-content {
            flex: 0 0 40%; /* 固定图片宽度占40%，可以根据需要调整 */
            max-width: 400px; /* 限制图片最大宽度 */
        }

        .image-content img {
            width: 100%; /* 图片宽度自适应 */
            height: auto; /* 保持图片比例 */
            border-radius: 8px; /* 圆角 */
            box-shadow: 0 2px 4px rgba(0, 0, 0, 0.1); /* 添加阴影 */
            border: 2px solid #ccc; /* 边框 */
        }
    </style>


    <img class="header-image" src="figures/S11_1.jpg" alt="Interview Image">
    <div class="article-link">
        <p>The Article Link:</p><a href="https://www.cell.com/cell/fulltext/S0092-8674(24)01268-6" target="_blank">Atlas of the plasma proteome in health and disease in 53,026 adults</a>
    </div>
    <div class="interviewee-introduction">
    <img src="figures/Jianfeng_Feng.jpg" alt="Interviewee Name" class="interviewee-photo">
    <div class="interviewee-details">
        <h2 class="anchored">Dr. Jianfeng Feng</h2>
        <p>Dr. Jianfeng Feng is the chair professor of Shanghai National Centre for Mathematic Sciences, the Dean of Brain-inspired AI Institute in Fudan University, and the dean of the School of Data Science. He has made considerable contributions on developing brain-inspired AI algorithms and applying them to tackle challenges raised in neuroscience and mental health with many publications in Cell, Nature Medicine, Nature Human Behaviour, Nature Aging, Nature Mental Health etc. He led a team to implement for the first time in the world the digital twin brain which has 86B neurons and 100T parameters. He was awarded the 2023 Humboldt Research Award, prestigious Royal Society Wolfson Research Merit Award, and invited to deliver 2019 Paykel Lecture at the Cambridge University.
        </p>
        </div>
    </div>
    
    <div class="interviewee-introduction">
    <img src="figures/Jintai_Yu.jpg" alt="Interviewee Name" class="interviewee-photo">
    <div class="interviewee-details">
        <h2 class="anchored">Dr. Jintai Yu</h2>
        <p>Dr. Jintai Yu is the Deputy Head of Department of Neurology in Huashan Hospital and the Vice Director of the Institute of Neurology, WHO Collaborating Center for Research and Training in Neurosciences, Fudan University, Shanghai, China. He also directs the Memory Disorders Subspecialty, National Center for Neurological Disorders in China, Huashan Hospital. Currently, he is focusing on basic and clinical research for Alzheimer’s disease and related dementia, and is the PI of several national grants on dementia, including Science and Technology Innovation 2030 Major Projects and National Natural Science Foundation of China. His research interests focus on the prevention, diagnosis, and treatment of AD, with the aim of applying research findings to clinical practice.
        </p>
        </div>
    </div>
        
        <div class="interview">
            <div class="answer-section">
                <p class="question-title">1. Regarding the research background and significance, does this work discover new knowledge or solve existing problems within the field? Please elaborate in detail.</p>
                <div class="answer-paragraph">
                    Large-scale proteomic studies can refine the understanding of health and disease. Nevertheless, most current proteomic studies have only focused on limited disease outcomes. The lack of a comprehensive human proteome-phenome atlas raises many questions. For example, are the associated proteins specific or shared among diseases and health-related traits, and can those proteomic profiles facilitate biological classification of human diseases? How do those plasma proteins contribute to minimally invasive assessment and tracking of hundreds of diseases? Are those associated proteins causally related to diseases, and do those causal proteins hold the potential to be therapeutic targets?

                </div>
                <div class="answer-paragraph">
                    This work present a comprehensive atlas of proteome-phenome associations (<a href="https://proteome-phenome-atlas.com/" target="_blank">https://proteome-phenome-atlas.com/</a>) by systematically mapping 2,920 plasma proteins to the presence and onset of 720 diseases and 986 health-related traits in 53,026 individuals. This atlas provides insights into shared and characteristic biological mechanisms among diseases. The proteomic profiles, coupled with machine learning, identify useful biomarkers and prediction models for multiple health conditions simultaneously. Through integrating protein quantitative trait locus (pQTL) data, this work also illustrates the use of the atlas for causal protein discovery and further drug target prioritization. The proteome-phenome atlas furnishes an extensive resource supporting future research in screening, diagnosis, and treatment of human diseases and advancing precision medicine.
                    <img class="inline-image" src="figures/S11_0.jpg" alt="Interview Image">
                </div>
            </div>
            <div class="answer-section">
                <p class="question-title">2. How did the reviewers evaluate (praise) it?</p>
                <div class="answer-paragraph">
                    Reviewers found this a comprehensive resource that holds promise for advancing precision medicine research. This paper presents an impressive amount of data that is of huge value for the scientific community. A particular strength of this study is the integration with pQTL and diseases genome-wide association analysis (GWAS) data to perform proteome-by-phenome Mendelian randomization. This provides insights into proteins causally linked to disease, which is critically important to develop treatments.
                </div>
            </div>
            
            
            <div class="answer-section">
                <p class="question-title">3. If this achievement has potential applications, what are some specific applications it might have in a few years?</p>
                <div class="answer-paragraph">
                    This work provides an open-access proteome-phenome resource (https://proteome-phenome-atlas.com/) revealing 168,100 protein-disease associations and 554,488 protein-trait associations. Moving forward, the research community will benefit from this comprehensive human proteome-phenome atlas to elucidate the biological mechanisms of diseases and accelerate the development of disease biomarkers.
                </div>
                <div class="answer-paragraph">
                    Integrating pQTL with diseases GWAS data, this work determines 474 potential causal protein-disease pairs using proteome-by-phenome Mendelian randomization, providing 37 drug repurposing opportunities and prioritizing 26 novel targets with favorable safety profiles. These findings may provide directive significance for drug repositioning and future drug development.
                </div>
            </div>
            
            <div class="answer-section">
                <p class="question-title">4. Can you recount the specific steps or stages from setting the research topic to the successful completion of the research?</p>
                <div class="answer-paragraph">
                    The initial attempt is our previous work, a proteomics study on Alzheimer’s disease (AD), that identified valuable potential biomarkers for risk prediction (DOI: 10.1038/s43587-023-00565-0). This encouraged us to explore the value of proteins for human health. We further conducted literature review and found existing gaps and societal needs, particularly regarding whether the identified proteins are AD-specific or shared across other diseases, and whether these proteins could serve as potential biomarkers or therapeutic targets in broader contexts. This led to the formation of our research topic: creating a comprehensive protein atlas for diseases and health-related traits, to answer the research questions (as detailed in Q1).
                </div>
                <div class="answer-paragraph">
                    Next, we defined the phenotypes and population. In order to accurately define the diseases, we used ICD-10 codes and conducted rigorous quality control. Considering that protein alterations could occur before or after disease onset, we treated diseases as both prevalent and incident. Health-related traits were also incorporated as critical disease indicators. Then, with our research design in place, we leveraged the expertise of our interdisciplinary team to select the optimal algorithms. We then conducted the analysis, followed by writing and peer review, culminating in the successful completion of the project.
                </div>
            </div>
            
            <div class="answer-section">
                <p class="question-title">5. Were there any memorable events during the research? You can tell a story about anything related to people, events, or objects.</p>
                <div class="answer-paragraph">
                    When reflecting on memorable events during the research, one particular moment stands out. As we were clustering the 660 diseases into 40 clusters, we found that the diseases from different chapters were gathered around, some of which were diseases with shared etiology. It was one of those moments where the potential implications of our work became much more exciting. Everyone, from data scientists to clinical researchers, began contributing ideas on how to further interpret these disease clusters. After the conversation, we conducted further investigation on the characteristic pathways and the comorbity status, which helped us to refine our research questions and expand the scope of the project. We archieved a real collaboration where the collective expertise of the team came together to push the boundaries of our initial vision. It was a reminder that research is not just about data—it’s about the ideas and interactions that propel us forward.
                </div>
            </div>
            
            <div class="answer-section">
                <p class="question-title">6.  Is there a follow-up plan based on this research? If so, please elaborate.</p>
                <div class="answer-paragraph">
                    Yes, there is a clear follow-up plan based on the findings. First, we plan to validate the protein biomarkers in clinical practice, with a particular focus on minimal-invasive approaches like blood-based biomarkers. Furthermore, we are considering collaborations with pharmaceutical companies to explore the therapeutic potential of the identified proteins, particularly those that show promise as druggable targets.
                </div>
            </div>
            
            <div class="answer-section">
                <p class="question-title">7. Without a doubt, AI is one of the hot topics of 2024, requiring extensive data support in its development. What assistance can biostatistics offer to the development of AI?</p>
                <div class="answer-paragraph">
                    AI is largely grounded in data-driven algorithms and applications, whereas biostatistics delves deeper into theoretical frameworks. Biostatistics offers invaluable methodologies to derive meaningful and persuasive solutions tailored to the unique datasets encountered in research, while AI presents a more straightforward approach to addressing complex queries. In my point, statisticians should try to jump out of their traditional boundaries and embrace the realm of various big data, and AI developers should pay more attention on mathematical foundations. In the age of big data, no data scientist should overlook the fundamental challenges inherent in the data itself.
                </div>
            </div>
            
            <div class="answer-section">
                <p class="question-title">8. Besides the above questions, is there anything else about this achievement that you would like to add? If so, please add it below.</p>
                <div class="answer-paragraph">
                    Another key takeaway from this work is the validation of the hypothesis that omics data, when combined with advanced AI models, can significantly enhance disease diagnosis, risk prediction, and therapeutic development. The ability to integrate omics data such as genomic, proteomic, metabolmic, and clinical data into cohesive models will further facilitate precision medicine.
                </div>
            </div>
            
        </div>
        
        <div class="edit-info">
            Edited by: Shan Gao<br>
            Proofread by: Hongtu Zhu
        </div>
        
<div style="display: flex; align-items: center; color: #8E8E8E; font-size: 13px; font-family: Arial, sans-serif;">Page Views: <div class="statcounter" style="margin-left: 10px;"> <!-- Add this line for styling -->
<!-- Default Statcounter code for s11
https://statsupai.org/quarto_web/site/posts/S10_interview.html
-->
<script type="text/javascript">
var sc_project=13068110; 
var sc_invisible=0; 
var sc_security="a70f159f"; 
var scJsHost = "https://";
document.write("<sc"+"ript type='text/javascript' src='" +
scJsHost+
"statcounter.com/counter/counter.js'></"+"script>");
</script>
<noscript><div class="statcounter"><a title="Web Analytics" href="https://statcounter.com/" target="_blank"><img class="statcounter" src="https://c.statcounter.com/13068110/0/a70f159f/0/" alt="Web Analytics" referrerpolicy="no-referrer-when-downgrade"></a></div></noscript>
<!-- End of Statcounter Code -->







</div></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>