<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="dcterms.date" content="2025-03-29">

<title>Mining the Unseen: How iKraph Transforms Literature into Actionable Biomedical Intelligence</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../styles.css">
</head>

<body>

<div id="quarto-search-results"></div>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
        
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Mining the Unseen: How iKraph Transforms Literature into Actionable Biomedical Intelligence</h1>
  <div class="quarto-categories">
    <div class="quarto-category">Interviews</div>
  </div>
  </div>



<div class="quarto-title-meta">

    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">March 29, 2025</p>
    </div>
  </div>
  
    
  </div>
  


</header>





    <link href="https://fonts.googleapis.com/css2?family=Roboto:wght@400;700&amp;display=swap" rel="stylesheet">

    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Mining the Unseen: How iKraph Transforms Literature into Actionable Biomedical Intelligence</title>

    <!-- Open Graph / Facebook -->
    <meta property="og:type" content="article">
    <meta property="og:url" content="https://statsupai.com/pages/S12.html">
    <meta property="og:title" content="Mining the Unseen: How iKraph Transforms Literature into Actionable Biomedical Intelligence">
    <meta property="og:description" content="">
    <meta property="og:image" content="figures/ikgraph.png">
    <style>
        body {
            font-family: 'Roboto', sans-serif;
            margin: 0;
            padding: 0;
            background-color: #f4f4f4;
            line-height: 1.6;
        }
        .interviewee-introduction {
            display: flex;
            align-items: center;
            background-color: #f8f8f8; /* 轻灰色背景 */
            border-radius: 8px; /* 边角圆润 */
            padding: 20px;
            margin-bottom: 20px; /* 和问答部分保持一定距离 */
        }

        .interviewee-photo {
            width: 150px; /* 根据需要调整 */
            height: 150px; /* 根据需要调整 */
            border-radius: 50%; /* 创建圆形图片 */
            object-fit: cover; /* 确保图片内容适应尺寸 */
            margin-right: 20px;
        }

        .interviewee-details h2 {
            margin-top: 0;
            color: #082653; /* 深蓝色，你可以根据网站主题更改 */
        }

        .interviewee-details p {
            color: #333; /* 深灰色文本 */
            line-height: 1.6; /* 增加行距 */
        }

        .header-image {
            width: 100%; /* 设置宽度为文本宽度的 80% */
            max-width: 1400px; /* 设置图片的最大宽度，与文本区域匹配 */
            height: auto; /* 确保图片比例不失真 */
            display: block; /* 使图片成为块级元素 */
            margin: 20px auto; /* 居中对齐，增加上下间距 */
            border-radius: 8px; /* 圆角处理，增加柔和感 */
            box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1); /* 轻微阴影，减轻视觉重量 */
            border: 3px solid #ddd; /* 边框颜色与页面风格统一 */
        }
        .inline-image {
            width: 50%; /* 设置宽度为容器宽度的一半，可根据需要调整 */
            max-width: 600px; /* 限制图片最大宽度 */
            height: auto; /* 保持图片比例 */
            display: block; /* 将图片作为块级元素显示 */
            margin: 10px auto; /* 居中对齐，增加间距 */
            border-radius: 4px; /* 圆角处理 */
            box-shadow: 0 2px 4px rgba(0, 0, 0, 0.1); /* 添加轻微阴影 */
            border: 2px solid #ccc; /* 边框颜色柔和，与文字风格一致 */
        }

        .interview {
            margin: 20px;
            background-color: rgba(255, 255, 255, 0.9);
            padding: 20px;
            border-radius: 8px;
            box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1);
        }

        .answer-section {
            margin-bottom: 30px;
            padding-bottom: 20px;
            border-bottom: 2px solid #ccc;
        }

        .question-title {
            font-size: 1.2rem;
            color: #003366;
            font-weight: bold;
            margin-bottom: 15px;
        }

        .answer-paragraph {
            margin-bottom: 15px;
            background-color: #f9f9f9;
            padding: 10px;
            border-radius: 5px;
            line-height: 1.8;
            box-shadow: 0 2px 4px rgba(0, 0, 0, 0.1);
        }

        .answer-paragraph:last-child {
            margin-bottom: 0;
        }

        .edit-info {
            color: #808080;
            font-size: 15px;
            text-align: right;
            padding: 20px 0;
            margin-top: 20px;
            border-top: 1px solid #ccc;
        }

        .article-link {
            background-color: #f4f4f4;
            padding: 10px;
            border-radius: 8px;
            text-align: center;
            margin: 20px 0;
        }

        .article-link p {
            margin: 0;
            color: #003366;
            font-weight: bold;
        }

        .article-link a {
            display: inline-block;
            background-color: #003366;
            color: white;
            padding: 10px 20px;
            border-radius: 5px;
            text-decoration: none;
            font-size: 18px;
            margin-top: 10px;
            transition: background-color 0.3s;
        }

        .article-link a:hover {
            background-color: #0056b3;
        }
        
        .text-image-container {
            display: flex;
            align-items: flex-start; /* 对齐方式 */
            gap: 20px; /* 设置文字和图片之间的间距 */
            margin: 20px 0; /* 增加上下间距 */
        }

        .text-content {
            flex: 1; /* 占用容器的一部分，自动调整宽度 */
        }

        .image-content {
            flex: 0 0 40%; /* 固定图片宽度占40%，可以根据需要调整 */
            max-width: 400px; /* 限制图片最大宽度 */
        }

        .image-content img {
            width: 100%; /* 图片宽度自适应 */
            height: auto; /* 保持图片比例 */
            border-radius: 8px; /* 圆角 */
            box-shadow: 0 2px 4px rgba(0, 0, 0, 0.1); /* 添加阴影 */
            border: 2px solid #ccc; /* 边框 */
        }
        .answer-paragraph p {
                margin-bottom: 1em;
            }
        .highlight {
                color: #007BFF; /* 重点色，蓝色 */
                font-weight: bold;
            }
        .section-title {
                font-weight: bold;
                font-size: 1.1em;
            }
    </style>


    <img class="header-image" src="figures/ikgraph.png" alt="Interview Image">
    <div class="article-link">
        <p>The Article Link:</p><a href="https://www.nature.com/articles/s42256-025-01014-w" target="_blank">A comprehensive large-scale biomedical knowledge graph for AI-powered data-driven biomedical research</a>
    </div>
    <div class="interviewee-introduction">
    <img src="figures/ikgraph1.jpg" alt="Interviewee Name" class="interviewee-photo">
    <div class="interviewee-details">
        <h2 class="anchored">Dr. Yuan Zhang</h2>
        <p>Dr. Yuan Zhang, is the Chief Scientist of R&amp;D at Insilicom LLC, leading the AI team in developing advanced AI models for biomedical research. With expertise in natural language processing, knowledge graph construction, and AI-driven drug discovery, Dr. Zhang has played a key role in building iKraph, a large-scale biomedical knowledge graph, and developing algorithms for adverse drug event monitoring and drug repurposing. Their work integrates AI, probabilistic semantic reasoning, and graph-based models to accelerate scientific discovery in the biomedical field.
        </p>
        </div>
    </div>
    
    <div class="interviewee-introduction">
    <img src="figures/ikgraph2.jpg" alt="Interviewee Name" class="interviewee-photo">
    <div class="interviewee-details">
        <h2 class="anchored">Dr. Xin Sui</h2>
        <p>Dr. Xin Sui, researches Large Language Model applications in biomedical text mining at FSU. With a PhD in Statistics, he specializes in biomedical text mining and knowledge graph construction. A LitCoin NLP Challenge winner, his work focuses on extracting knowledge from scientific literature.
        </p>
        </div>
    </div>
    
    <div class="interviewee-introduction">
    <img src="figures/ikgraph3.jpg" alt="Interviewee Name" class="interviewee-photo">
    <div class="interviewee-details">
        <h2 class="anchored">Dr. Feng Pan</h2>
        <p>Dr. Feng Pan, Associate Director of Data Science, Insilicom LLC: Expert in biomedical data
        processing, natural language processing and LLM prompting, with a proven track record as a key
        contributor to the LitCoin NLP Challenge and BioCreative competitions.
        </p>
        </div>
    </div>
    
    <div class="interviewee-introduction">
    <img src="figures/ikgraph4.jpg" alt="Interviewee Name" class="interviewee-photo">
    <div class="interviewee-details">
        <h2 class="anchored">Dr. Kaixian Yu</h2>
        <p>Kaixian Yu, PhD in Statistics, currently serving as CTO at Insilicom LLC, previously an expert researcher at Didi Global's AI Labs, head of the community e-commerce search and recommendation system, and head of data science for growth at Shopee. Dedicated to applying statistical methods to industrial practices, with research areas including statistical learning, natural language processing, computer vision, graph neural networks, reinforcement learning, and causal inference. Published several papers in top-tier journals such as Nature and Cell, and served as a program committee member for leading conferences including KDD, IJCAI, and AAAI.
        </p>
        </div>
    </div>
    
    <div class="interviewee-introduction">
    <img src="figures/ikgraph5.jpg" alt="Interviewee Name" class="interviewee-photo">
    <div class="interviewee-details">
        <h2 class="anchored">Dr. Jinfeng Zhang</h2>
        <p>Dr. Jinfeng Zhang obtained his bachelor’s degree from Peking University in Beijing, China in 1997. He then came to the US to pursue graduate studies at the University of Illinois at Chicago, where he obtained a master’s degree in chemistry (2001), master’s degree in mathematics and computer science (2002), and PhD in bioinformatics (2004). He then received postdoctoral training in the department of statistics at Harvard University during 2004-2007. He joined the department of statistics at Florida State University as a faculty member in 2007. His research encompasses applications of computational methods and artificial intelligence in biology, including biological information extraction, natural language processing, biomedical knowledge graphs, and AI for Science. In the summer of 2022, Dr. Zhang took a leave of absence from Florida State University to dedicate his time fully to his startup company, Insilicom LLC.
        </p>
        </div>
    </div>
        
        <div class="interview">
            <div class="answer-section">
                <p class="question-title">1. Regarding the research background and significance, does this work discover new knowledge or solve existing problems within the field? Please elaborate in detail.</p>
                <div class="answer-paragraph">
                    <p>
                        The work both uncovers new knowledge and addresses longstanding challenges in biomedical research. Here are multiple perspectives:
                    </p>
                    <p class="section-title">
                        1. Novel Discovery of Knowledge
                    </p>
                    <p>
                        The paper introduces iKraph—a knowledge graph constructed by processing all PubMed abstracts with a pipeline that achieves human-level accuracy. This effort goes beyond traditional manual curation by integrating 10+ million entities and over 30 million relations, thereby revealing novel associations and insights not captured in existing public databases. By aggregating and interpreting vast amounts of literature, the method uncovers previously hidden or underappreciated relationships (e.g., between diseases, genes, and drugs) that can inspire new hypotheses and research directions.
                    </p>
                    <p class="section-title">
                        2. Solving Existing Methodological Problems
                    </p>
                    <p>
                        The study addresses a critical bottleneck in biomedical informatics: the extraction of structured, reliable knowledge from unstructured text. Traditional approaches struggled with scale and accuracy, but this work leverages advanced natural language processing and a winning pipeline from the LitCoin NLP Challenge to meet these challenges head-on. The development of the probabilistic semantic reasoning (PSR) algorithm is particularly noteworthy—it provides an interpretable framework for inferring indirect causal relationships, a task that has proven challenging with earlier methods. This directly tackles the problem of making sense of complex, multi-hop relations in biomedical data.
                    </p>

                    <p class="section-title">
                        3. Impact on Applications like Drug Repurposing
                    </p>
                    <p>
                        The integration of diverse data sources and the ability to infer causal pathways allow the system to generate actionable insights, as demonstrated by its application in real-time drug repurposing studies (e.g., for COVID-19 and cystic fibrosis). This not only speeds up hypothesis generation but also improves the rigor of validation compared to traditional approaches. The work shows that structured, high-quality knowledge graphs can be directly used to predict and validate novel therapeutic candidates—bridging a gap that manual literature reviews and less sophisticated automated systems have struggled to close.
                    </p>

                    <p>
                        Overall, the study exemplifies a paradigm shift: it both discovers new biomedical insights by mining and integrating massive amounts of data, and it provides robust solutions to longstanding challenges in automated knowledge extraction and hypothesis generation.
                    </p>
                </div>
            </div>
            
            <div class="answer-section">
                <p class="question-title">2. How did the reviewers evaluate (praise) it?</p>
                <div class="answer-paragraph">
                    <p>
                        The reviewers were generally very positive about the work, highlighting several aspects that set it apart:
                    </p>
                    <p class="section-title">
                        1. Scale and Integration
                    </p>
                    <p>
                        They praised the sheer scale of the knowledge extraction—processing over 34 million PubMed abstracts and integrating data from 40 public databases and high-throughput genomics analyses—which provides a much richer and more comprehensive resource than previous efforts. The integration of diverse data sources was seen as a major strength, establishing a solid foundation for advanced biomedical insights.
                    </p>
                    <p class="section-title">
                        2. Methodological Advances and Interpretability
                    </p>
                    <p>
                        Both reviewers commended the use of an advanced information extraction pipeline (the LitCoin Challenge-winning system) that achieves human-level performance. The development of the probabilistic semantic reasoning (PSR) framework was especially highlighted for its interpretability. Unlike many “black-box” models, PSR allows experts to trace the reasoning behind inferred indirect relationships, which is critical for applications such as drug repurposing.
                    </p>
                    <p class="section-title">
                        3. Impact on Drug Repurposing and Therapeutic Discovery
                    </p>
                    <p>
                        The system’s performance in drug repurposing studies—demonstrating high recall rates in case studies for COVID-19 and cystic fibrosis—was seen as compelling evidence of its practical utility in accelerating therapeutic discovery. The retrospective validation showing that many viable drug candidates could have been identified early (for example, during the COVID-19 pandemic) impressed the reviewers, suggesting real-world impact.
                    </p>
                    <p class="section-title">
                        4. Overall Contribution to Biomedical Research
                    </p>
                    <p>
                        Reviewer #1 noted that while no single element might be entirely novel, the integrated system as a whole represents a significant advancement in automated biomedical knowledge extraction and inference. Reviewer #2 emphasized the KG’s value for the biomedical community, especially given that many existing KGs are outdated, and appreciated the efforts to provide both a server interface and downloadable data.
                    </p>
                    <p>
                        In summary, the reviewers evaluated the work as a credible and meaningful step forward in biomedical informatics—combining scale, methodological rigor, and interpretability to deliver a resource with clear applications in accelerating research and drug discovery.
                    </p>
                </div>
            </div>
            
            
            <div class="answer-section">
                <p class="question-title">3. If this achievement has potential applications, what are some specific applications it might have in a few years?</p>
                <div class="answer-paragraph">
                    <p>
                        The achievement paves the way for several transformative applications in biomedical research and healthcare. For instance:
                    </p>
                    <p class="section-title">
                        1. Drug Repurposing and Therapeutic Discovery
                    </p>
                    <p>
                        The system’s ability to infer indirect causal relations can be further refined to rapidly identify novel drug candidates or repurpose existing ones, particularly for emerging diseases. Future iterations could integrate real-time data feeds to continuously update predictions, enabling faster response during health crises.
                    </p>
                    <p class="section-title">
                        2. Precision Medicine and Patient Stratification
                    </p>
                    <p>
                        With a comprehensive view of gene-disease-drug associations, clinicians could better tailor treatments to individual patients by identifying molecular targets specific to a patient’s condition. The knowledge graph could be integrated with patient genomic data to predict treatment responses and potential side effects.
                    </p>
                    <p class="section-title">
                        3. Automated Hypothesis Generation for Research
                    </p>
                    <p>
                        Researchers can use the system as an automated assistant to generate and prioritize new scientific hypotheses, highlighting underexplored biological pathways or novel interactions. This could accelerate basic research and drive the discovery of previously unknown mechanisms of disease.
                    </p>
                    <p class="section-title">
                        4. Enhanced Clinical Decision Support
                    </p>
                    <p>
                        The KG can serve as a backbone for decision support systems in clinical settings, providing clinicians with evidence-backed insights drawn from vast amounts of literature. For example, it could aid in diagnosing complex conditions by linking clinical symptoms with underlying molecular mechanisms.
                    </p>
                    <p class="section-title">
                        5. Integration with Emerging AI Models
                    </p>
                    <p>
                        As large language models continue to evolve, combining them with a robust, data-driven KG like iKraph could enhance the accuracy and interpretability of biomedical AI tools, leading to smarter clinical and research applications.
                    </p>
                    <p>
                        These applications underscore the long-term potential of this achievement to not only transform biomedical research but also to directly impact patient care and healthcare innovation.
                    </p>
                </div>
            </div>
            
            <div class="answer-section">
                <p class="question-title">4. Can you recount the specific steps or stages from setting the research topic to the successful completion of the research?</p>
                <div class="answer-paragraph">
                    <p class="section-title">
                        1. Historical Foundation and Early Vision
                    </p>
                    <p>
                        We began with a long-held vision of using knowledge graphs for automated knowledge discovery. Our journey started in 2011 with our PLOS One paper, <a href="https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0021474" class="highlight">
                            “Integrated Bio-Entity Network: A System for Biological Knowledge Discovery”
                        </a>, which laid the conceptual groundwork for today’s approach—even though back then we did not use the term “knowledge graph.”
                    </p>
                    <p>
                        At that time, our extraction methods were more rudimentary, and the quality of the information extracted was not comparable to what is achievable now.
                    </p>
                    <p class="section-title">
                        2. Advancements in Technology and Methodology
                    </p>
                    <p>
                        With the emergence of deep learning and large language models (LLMs), we revamped our approach by developing an award-winning, human-level information extraction pipeline. This new system could accurately extract multiple critical biological entities and their relationships from vast datasets, enabling us to construct a comprehensive, high-quality knowledge graph.
                    </p>
                    <p>
                        Recognizing that understanding causal directions is key for generating actionable hypotheses, we took an extra step to manually annotate causal information and build a causal knowledge graph.
                    </p>
                    <p class="section-title">
                        3. Algorithm Development and Validation
                    </p>
                    <p>
                        A pivotal stage was the creation of our probabilistic semantic reasoning (PSR) algorithm. This probabilistic framework allowed us to infer indirect causal relationships at scale, generating interpretable hypotheses from our graph.
                    </p>
                    <p>
                        We also implemented a time-sensitive validation strategy. By retrospectively testing our generated hypotheses against evolving data, we rigorously demonstrated the effectiveness and robustness of our overall approach.
                    </p>
                    <p>
                        Each of these stages—from the initial conceptualization to technological enhancement, algorithm development, and rigorous validation—played an essential role in the successful completion of the research.
                    </p>
                </div>
            </div>
            
            <div class="answer-section">
                <p class="question-title">5. Were there any memorable events during the research? You can tell a story about anything related to people, events, or objects.</p>
                <div class="answer-paragraph">
                    <p>
                        Our journey was full of ups and downs that ultimately enriched our research story:
                    </p>
                    <p>
                        We initially submitted the manuscript to top-tier journals like <span class="highlight">Nature</span> and <span class="highlight">Science</span>, only to face rejections without even reaching peer review—and <span class="highlight">Nature Biotechnology</span> also passed us by. That period was very disheartening, and for a while, we felt too frustrated to continue submitting to similar outlets.
                    </p>
                    <p>
                        However, rather than giving up, we decided to persevere and submitted the work to <span class="highlight">Nature Machine Intelligence</span>. This time, our efforts paid off, and the manuscript was accepted—a turning point that reaffirmed our belief in the value of our approach.
                    </p>
                    <p>
                        Along the way, one of the most memorable surprises was discovering that our method could repurpose a remarkably large number of drugs and diseases, each backed by substantial literature evidence. Reading through these supporting pieces of evidence made us realize the immense potential of our work.
                    </p>
                    <p>
                        At the same time, the process also revealed a critical insight: while our predictions were exciting, many repurposed drug candidates were not ideal due to simple, overlooked factors. This observation has already set the stage for our next phase of research, where reducing the false positive rate will be a key focus.
                    </p>
                    <p>
                        Earlier milestones, such as winning the <span class="highlight">LitCoin NLP Challenge</span> and <span class="highlight">BioCreative challenges</span>, also played pivotal roles in building our confidence and shaping the direction of our study. These successes, along with many precious memories from our collaborative efforts, continue to inspire us—even if I won’t elaborate on all those details here.
                    </p>
                    <p>
                        This rollercoaster of rejections, surprising discoveries, and key milestones not only defined our research journey but also reinforced the resilience and creativity required to drive innovation in biomedical knowledge discovery.
                    </p>
                </div>
            </div>
            
            <div class="answer-section">
                <p class="question-title">6.  Is there a follow-up plan based on this research? If so, please elaborate.</p>
                <div class="answer-paragraph">
                    <p>
                        We are applying the framework to perform a large-scale drug repurposing effort for all rare diseases. We have selected a set of rare cancers and conducted experimental validation of repurposed compounds on some rare cancer cell lines and obtained very promising results.
                    </p>
                    <p>
                        Another line of research is to use the knowledge graph to develop advanced link prediction methods to predict new drug targets, new treatments for diseases, and hypothesis generation in general.
                    </p>
                    <p>
                        We are building a question-answering system by combining a LLM with our KG and are also integrating knowledge from analyzing large volumes of public genomics data.
                    </p>
                </div>
            </div>
            
            <div class="answer-section">
                <p class="question-title">7. Without a doubt, AI is one of the hot topics of 2025, requiring extensive data support in its development. What assistance can biostatistics offer to the development of AI?</p>
                <div class="answer-paragraph">
                    <p>
                        <span class="highlight">Biostatistics</span> plays a pivotal role in powering AI developments, especially in data-intensive fields like biomedicine. Here are several ways it can contribute:
                    </p>
                    <p>
                        <span class="section-title">1. Robust Data Analysis and Quality Control.</span> Biostatistics provides the rigorous methodologies needed to clean, integrate, and interpret complex datasets. This is crucial for ensuring that AI models are built on reliable and high-quality data.
                    </p>
                    <p>
                        <span class="section-title">2. Experimental Design and Sample Size Estimation.</span> In designing studies or experiments for AI research, biostatistical methods help determine optimal sample sizes and experimental protocols. This ensures that the collected data is both representative and statistically robust, reducing bias and enhancing model generalizability.
                    </p>
                    <p>
                        <span class="section-title">3. Uncertainty Quantification and Model Validation.</span> Biostatistics offers techniques for quantifying uncertainty, which are essential for assessing the reliability of AI predictions. By incorporating confidence intervals, hypothesis testing, and significance testing, researchers can better validate their models and results.
                    </p>
                    <p>
                        <span class="section-title">4. Causal Inference and Interpretation.</span> Many AI applications, especially in healthcare, require not just correlation but a deep understanding of causality. Biostatistical tools enable causal inference, allowing researchers to identify and validate the underlying relationships between variables—a key aspect for interpreting AI outcomes.
                    </p>
                    <p>
                        <span class="section-title">5. Development of Predictive Models.</span> Statistical models such as logistic regression, survival analysis, and Bayesian inference provide foundational tools that have been extended into modern AI. These methods help in the development and refinement of predictive models, serving as a bridge between traditional statistical methods and cutting-edge machine learning techniques.
                    </p>
                    <p>
                        <span class="section-title">6. Evaluation and Optimization of AI Systems.</span> Biostatistics contributes to the evaluation of AI systems through the development of performance metrics and validation frameworks. This ensures that AI models are not only accurate but also robust and reproducible across different datasets and conditions.
                    </p>
                    <p>
                        In summary, biostatistics brings rigorous quantitative methods and a deep understanding of data variability and uncertainty, which are indispensable for the development, validation, and interpretation of AI systems.
                    </p>
                </div>
            </div>
            
        </div>
        
        <div class="edit-info">
            Edited by: Shan Gao<br>
            Proofread by: Hongtu Zhu
        </div>
        
<div style="display: flex; align-items: center; color: #8E8E8E; font-size: 13px; font-family: Arial, sans-serif;">Page Views: <div class="statcounter" style="margin-left: 10px;"> <!-- Add this line for styling -->
<!-- Default Statcounter code for s12
https://statsupai.org/pages/S12.html -->
<script type="text/javascript">
var sc_project=13109864; 
var sc_invisible=0; 
var sc_security="82300fc7"; 
var scJsHost = "https://";
document.write("<sc"+"ript type='text/javascript' src='" +
scJsHost+
"statcounter.com/counter/counter.js'></"+"script>");
</script>
<noscript><div class="statcounter"><a title="Web Analytics" href="https://statcounter.com/" target="_blank"><img class="statcounter" src="https://c.statcounter.com/13109864/0/82300fc7/0/" alt="Web Analytics" referrerpolicy="no-referrer-when-downgrade"></a></div></noscript>
<!-- End of Statcounter Code -->








</div></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>