<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.554">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="dcterms.date" content="2024-06-19">

<title>MaxFuse: Unleashing the Full Potential of Single Cell and Spatial Genomics Integration</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../styles.css">
</head>

<body>

<div id="quarto-search-results"></div>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">MaxFuse: Unleashing the Full Potential of Single Cell and Spatial Genomics Integration</h1>
  <div class="quarto-categories">
    <div class="quarto-category">Interviews</div>
  </div>
  </div>



<div class="quarto-title-meta">

    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">June 19, 2024</p>
    </div>
  </div>
  
    
  </div>
  


</header>





    <link href="https://fonts.googleapis.com/css2?family=Roboto:wght@400;700&amp;display=swap" rel="stylesheet">

    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>MaxFuse: Unleashing the Full Potential of Single Cell and Spatial Genomics Integration</title>

    <!-- Open Graph / Facebook -->
    <meta property="og:type" content="article">
    <meta property="og:url" content="https://statsupai.org/pages/S7_interview.html">
    <meta property="og:title" content="MaxFuse: Unleashing the Full Potential of Single Cell and Spatial Genomics Integration">
    <meta property="og:description" content="">
    <meta property="og:image" content="figures/S7_1.jpg">
    <style>
        body {
            font-family: 'Roboto', sans-serif; 
            margin: 0;
            padding: 0;
            background-color: #f4f4f4;
            line-height: 1.6; /* 增加行距 */
        }
        .large-text {
            font-size: larger; /* 或你想要的任何大小 */
        }

        .interviewee-introduction {
            display: flex;
            align-items: center;
            background-color: #f8f8f8; /* 轻灰色背景 */
            border-radius: 8px; /* 边角圆润 */
            padding: 20px;
            margin-bottom: 20px; /* 和问答部分保持一定距离 */
        }

        .interviewee-photo {
            width: 150px; /* 根据需要调整 */
            height: 150px; /* 根据需要调整 */
            border-radius: 50%; /* 创建圆形图片 */
            object-fit: cover; /* 确保图片内容适应尺寸 */
            margin-right: 20px;
        }

        .interviewee-details h2 {
            margin-top: 0;
            color: #082653; /* 深蓝色，你可以根据网站主题更改 */
        }

        .interviewee-details p {
            color: #333; /* 深灰色文本 */
            line-height: 1.6; /* 增加行距 */
        }
        .edit-info {
            color: #808080; /* 设置为灰色字体 */
            font-size: 15px; /* 设置字体大小，根据需要调整 */
            text-align: right; /* 右对齐文本 */
            padding: 20px 0; /* 在内容上下添加一些空间 */
            margin-top: 20px; /* 与问答内容保持一定距离 */
            border-top: 1px solid #ccc; /* 如果需要，可以添加一条分隔线 */
        }

        .header-image {
            width: 97%;
            height: auto;
            display: block; /* 使图片成为块级元素 */
            margin: 0 auto; /* 水平居中 */
        }
        .fixed-counter {
            position: fixed; /* 固定定位，使元素不随页面滚动 */
            left: 0; /* 距离页面左边界的距离 */
            bottom: 0; /* 距离页面底部边界的距离 */
            background-color: transparent; /* 可以根据需要设置背景颜色 */
            z-index: 1000; /* 确保计数器显示在其他元素之上 */
        }

        .interview {
        margin: 20px;
        background-color: rgba(255, 255, 255, 0.7); /* 白色背景，70%不透明度（即30%透明度） */
        padding: 20px;
        border-radius: 8px;
        box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1);
        }

        /* 标题和子标题 */
        h1, h2 {
            font-weight: bold;
            color: #7469B6; /* 深蓝色 */
            margin-bottom: 20px;
        }

        /* 问题样式 */
        .question {
            color: #213c64; /* 深蓝色 */
            font-weight: bold;
            margin-bottom: 10px; /* 问题与回答之间的空间 */
            font-size: larger; /* 突出问题 */
        }
        .image {
            display: block; /* 使图像成为块级元素 */
            max-width: 60%; /* 设置图像的最大宽度为60%以确保响应式 */
            height: auto; /* 设置图像的高度自动调整以保持图片比例 */
            margin: 10px auto; /* 上下边距为20px，左右自动，实现水平居中 */
            border: 5px solid #808080; /* 添加蓝色的边框，宽度为5px */
            border-radius: 8px; /* 如果您想要圆角边框，可以添加这个属性 */
        }

        .answer {
            background-color: #f8f8f8; /* 轻微的背景色增加对比 */
            padding: 15px; /* 内边距增加可读性 */
            border-left: 4px solid #213c64; /* 左侧边框作为视觉提示 */
            margin-bottom: 20px; /* 增加回答间的空间 */
            color: #264653; /* 适中的文字颜色 */
            font-size: medium; /* 适合阅读的字号 */
            line-height: 1.6; /* 行间距增加可读性 */
            border-radius: 4px; /* 轻微的圆角 */
            box-shadow: 0 2px 4px rgba(0,0,0,0.1); /* 轻微的阴影增加层次感 */
        }
        ul.custom-list {
        padding-left: 20px; /* 调整以适应项目符号和首行缩进 */
        }
        ul.custom-list li {
            text-indent: 2em; /* 首行缩进 */
            margin-left: -2em; /* 调整项目符号位置 */
        }

        .background-image {
            position: fixed; /* 或absolute，取决于你的需求 */
            right: 0;
            bottom: 0;
            width: 800px; /* 根据你的背景图尺寸调整 */
            height: 300px; /* 根据你的背景图尺寸调整 */
            background-image: url('12.jpg'); /* 替换为你的图片名称 */
            background-size: cover;
            opacity: 0.4; /* 设置透明度为40% */
            z-index: -1; /* 确保背景图不会覆盖其他内容 */
        }
        .article-link {
            background-color: #f4f4f4; /* 背景色 */
            padding: 10px; /* 内边距 */
            border-radius: 8px; /* 边角圆滑 */
            text-align: center; /* 文本居中 */
            margin: 20px 0; /* 上下外边距 */
        }
    
        .article-link p {
            margin: 0; /* 移除默认边距 */
            color: #003366; /* 文本颜色 */
            font-weight: bold; /* 字体加粗 */
        }
    
        .article-link a {
            display: inline-block; /* 可以设置宽度和高度 */
            background-color: #213c64; /* 链接背景色 */
            color: white; /* 链接文字颜色 */
            padding: 10px 20px; /* 链接内边距 */
            border-radius: 5px; /* 链接边角圆滑 */
            text-decoration: none; /* 去掉下划线 */
            font-size: 18px; /* 字体大小 */
            margin-top: 10px; /* 与段落之间的距离 */
            transition: background-color 0.3s; /* 过渡动画 */
        }
    
        .article-link a:hover {
            background-color: #082653; /* 鼠标悬停时的背景色 */
        }
        

    </style>



    <img class="header-image" src="figures/S7_1.jpg" alt="Interview Image">

    <div class="article-link">
    <p>The Article Link:</p><a href="https://www.nature.com/articles/s41587-023-01935-0" target="_blank">Integration of spatial and single-cell data across modalities with weakly linked features</a>
    </div>

    <div class="interviewee-introduction">
    <img src="figures/S7_bio1.png" alt="Interviewee Name" class="interviewee-photo">
    <div class="interviewee-details">
        <h2 class="anchored">Dr. Zongming Ma</h2>
        <p>Dr. Ma is a Professor of Statistics and Data Science at Yale University. His current research interests include multi-modal data analysis, network data analysis, and statistical and machine learning methods in genomics and imaging.  

        Dr. Ma obtained his PhD in Statistics from Stanford University in 2010 and joined the Department of Statistics and Data Science at the Wharton School of the University of Pennsylvania as a tenure-track assistant professor in the same year. After working at the University of Pennsylvania for thirteen years, he moved to Yale University in 2023. He is the recipient of an NSF CAREER award in 2014, a Sloan Research Fellowship in Mathematics in 2016, and an Institute of Mathematical Statistics Fellowship in 2023.
        </p>
        </div>
    </div>
    <div class="interviewee-introduction">
    <img src="figures/S7_bio2.png" alt="Interviewee Name" class="interviewee-photo">
    <div class="interviewee-details">
        <h2 class="anchored">Dr. Nancy Ruonan Zhang</h2>
        <p>Dr. Zhang is a Ge Li and Ning Zhao Professor of Statistics in The Wharton School at University of Pennsylvania.  Her research focuses primarily on the development of statistical methods and computational algorithms for the analysis of data from high-throughput biological experiments.  She has made contributions to copy number and structural variant detection, to the modeling and estimation of intra-tumor genetic heterogeneity, and to the modeling and analysis of single-cell and spatial genomic data.  In Statistics, she has made contributions to change-point analysis, variable selection, and model selection.  

        Dr. Zhang obtained her Ph.D. in Statistics in 2005 from Stanford University.  After one year of postdoctoral training at University of California, Berkeley, she returned to the Department of Statistics at Stanford University as Assistant Professor in 2006.  She received the Sloan Fellowship in 2011, and formally moved to University of Pennsylvania with tenure in 2012.  She was awarded the Medallion Lectureship by the Institute of Mathematical Statistics in 2021.  Her work has been funded by grants from the NSF and NIH.  At Penn, she is a member of the Graduate Group in Genomics and Computational Biology, and currently serves as the Vice Dean of the Wharton Doctoral Program.
        </p>
        </div>
    </div>

    
    <div class="interview">
        <p class="question">Regarding the research background and significance, does this work discover new knowledge or solve existing problems within the field? Please elaborate in detail.</p>
        <p class="answer">The work solves existing problems within the field. To introduce the work, we need to first say a little bit about what is single cell and spatial genomics.  The last decade brought about a complete paradigm shift in biology where genome-wide measurements can be made on single cells, across many cells in parallel. By “genome-wide”, we mean, for example, measuring the expression of tens of thousands of genes at once, for each of thousands to millions of cells at once.  Such technologies are called “single cell technologies”.   Even more exciting, the past couple of years have  brought about “spatial genomic technologies”, which can make such measurements for cells in situ.  “In situ” means “fixed in space”, so that we know not only the cell’s internal molecular stage but also its tissue neighborhood.  Since cells form tight-knit communities and “collaborate” and “compete” with other cells in its tissue niche, spatial genomic technologies provide really important information for understanding cell and tissue biology.</p>

        <p class="answer">For such single cell and spatial data, a primary unsolved challenge is how to perform diagonal integration.  What is diagonal integration?   To thoroughly characterize a cell, we would like to measure the expression of all of its genes, as well as to quantify what proteins and metabolites it is making. We perhaps would also like to know the conformation of its chromatin (that is, which regions of its DNA are accessible, and which are “closed” and not available for transcription). Each of these modalities, on its own, provide valuable information but never the complete picture. Ideally, we would like to measure everything for each cell, but no technology is yet capable of that. Thus we are left with having to apply different technologies, each profiling a subset of the genes/modalities of interest.  Sometimes, for example, we have a spatial technology recording the tissue position of cells and their expression of a panel of proteins, and a single cell technology that does not provide the spatial coordinates but measures the full transcriptome.  In such cases, we would really like to have all three: spatial position, full transcriptome measurement, and protein measurement. The subsequent analysis challenge of “aligning” cells across technologies is often called diagonal integration.  Diagonal integration is extremely challenging when the data modalities are weakly linked, that is, when there are few or none shared features between data modalities.</p> 

        <p class="answer">In this paper, we provide a new method, MaxFuse, which, through iterative co-embedding, data smoothing and cell matching, uses all information in each modality to achieve high quality integration even in extremely challenging scenarios.  MaxFuse is, to our knowledge, the only method that can reliably integrate single cell RNA sequencing data with spatial proteomic data, which is a commonly appearing study design in current tissue atlasing efforts.</p>

        <p class="question">How did the reviewers evaluate (praise) it?</p>
        <p class="answer">The reviewers immediately recognized the importance of this problem and the impact of the method, but they made us do more benchmarking and exploration of model parameters before letting us publish it. We believe that the reviewers were very interested in how this method can standardize the computational pipelines for analysis of spatial proteomic and transcriptomic data.
        </p>


        <p class="question">If this achievement has potential applications, what are some specific applications it might have in a few years?</p>
        <p class="answer">As mentioned, single cell and spatial genome profiling technologies are now the bread-and-butter of biomedical research, but we are still at the early stages in realizing the potential of these technologies.  We believe MaxFuse can greatly increase the flexibility of study designs and allow for more comprehensive characterization of tissues through integration across data modalities. In addition, the design of the algorithm is modality agnostic, and so the method is very likely to continue to work for new modalities.</p>

        <p class="question">Can you recount the specific steps or stages from setting the research topic to the successful completion of the research?</p>
        <p class="answer">In applied problems, the initial step of defining the scope of the problem, and giving it a precise formulation, is always very important. We recognized that diagonal integration was an important unsolved problem, and the then existing methods being applied to this problem did not utilize anything except for the first moments of the shared features.  They did not work well except for cases where there are lots of overlapping features with strong correlation between modalities.  The correlation structures within each modality also contain valuable information about the geometry of the data, but were not used.  We wanted to use all of the information within each modality, not just the shared features, This, and other insights, such as the data denoising via graph smoothing, allows us to separate the signal from the noise. </p>

        <p class="answer">The project was also boosted by a collaboration with Garry Nolan’s lab, especially an extremely capable graduate student Bokai Zhu in the driver’s seat. As an interesting fact, the two co-first authors Shuxiao Chen and Bokai have been close friends since college, and the collaboration was originally initiated by them. Bokai brought a lot of insights about the different spatial technologies, and selected data sets to illustrate the power of the approach.  Often, for scientific papers, the selection of data sets is just as important as the development of the method.  We identified data sets that can be used for benchmarking (aka has known truth of how cells should be aligned) and data sets that can be used to tell an interesting story. These are extremely important and took up most of our time working on the paper.</p>

        <p class="question">Were there any memorable events during the research? You can tell a story about anything related to people, events, or objects.</p>
        <p class="answer">It is always memorable when with your own method you are able to uncover something new in data. We applied MaxFuse to the analysis of CODEX tonsil data. CODEX is a spatial technology that only aligns a small panel of proteins. We aligned the cells in CODEX to cells from scRNAseq of the same tissue, which enabled the precise definition of expression patterns of any gene in space. We then checked to see how the genes were distributed within B cell germinal centers, and found patterns that were biologically meaningful and made sense. This was not possible with any previous method.
        </p>
        

        <p class="question">Is there a follow-up plan based on this research? If so, please elaborate.</p>
        <p class="answer">We tried many different ideas and have other backup ideas up our sleeve in case things fail.  But we didn’t really deviate significantly from our original planned course.</p>

        <p class="question">Without a doubt, AI is one of the hot topics of 2023, requiring extensive data support in its development. What assistance can biostatistics offer to the development of AI?</p>
        <p class="answer">Although AI is a big exciting direction, it is not the best solution to everything.  To the problem of diagonal integration, AI has been applied, and a few have worked ok. However, none so far worked on the challenging cases of weak linkage, which motivated MaxFuse. 

        We showed in this paper that simpler but more thoughtful methods development can sometimes beat black box AI algorithms. As we have elaborated earlier, the statistical ideas of denoising, local smoothing, and canonical correlations have played important roles in MaxFuse.

        Successful applications of AI require understanding of the problem domain, and a good grasp of the signal and noise structures of the data.  In this aspect, successful application of AI is no different from successful application of Statistics.  We believe that what ever question we are addressing, the key to successful methods development is working closely with field scientists to understand the scientific problem and the data.
        </p>

        <!-- 添加更多问题和答案 -->

    </div>
    <div class="edit-info">
        Edited by: Shan Gao<br>
        Proofread by: Hongtu Zhu
    </div>
<div class="background-image"></div>
<!-- Default Statcounter code for Interview_peiwang
https://statsupai.com/pages/interview_weipeng.html -->
<div style="display: flex; align-items: center; color: #8E8E8E; font-size: 13px; font-family: Arial, sans-serif;">Page Views: <div class="statcounter" style="margin-left: 10px;"> <!-- Add this line for styling -->
<!-- Default Statcounter code for Interview_peiwang1
https://statsupai.com/pages/interview_weipeng.html -->
<script type="text/javascript">
var sc_project=12973370; 
var sc_invisible=0; 
var sc_security="a715d5f2"; 
var scJsHost = "https://";
document.write("<sc"+"ript type='text/javascript' src='" +
scJsHost+
"statcounter.com/counter/counter.js'></"+"script>");
</script>
<noscript><div class="statcounter"><a title="Web Analytics" href="https://statcounter.com/" target="_blank"><img class="statcounter" src="https://c.statcounter.com/12973370/0/a715d5f2/0/" alt="Web Analytics" referrerpolicy="no-referrer-when-downgrade"></a></div></noscript>
<!-- End of Statcounter Code -->





</div></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>