<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>Data Resources for Proteogenomics-related Studies</title>
    <style>
        body {
            font-family: 'Arial', sans-serif;
            line-height: 1.6;
            background-color: #f4f4f4;
            margin: 0;
            padding: 0;
        }
        .container {
            width: 85%;
            margin: auto;
            overflow: hidden;
            padding: 20px;
        }
        .header {
            background-color: #333;
            color: #fff;
            text-align: center;
            padding: 20px;
        }
        .card {
            background: #fff;
            box-shadow: 0 2px 4px rgba(0, 0, 0, 0.1);
            margin-bottom: 20px;
            padding: 20px;
            border-radius: 8px;
        }
        .card h2 {
            color: #333;
            margin-top: 0;
        }
        .card p {
            margin: 10px 0;
            font-size: 19px; /* Increased font size for descriptions */
        }
        .card a {
            display: inline-block;
            background: #003366; /* Deep blue background for links */
            color: #ffffff;
            padding: 8px 12px; /* Smaller padding for a slimmer look */
            border-radius: 5px;
            text-decoration: none;
            transition: background 0.3s;
            font-size: 16px; /* Adjusted font size for links */
        }
        .card a:hover {
            background: #002244; /* Darker blue on hover */
        }
        .intro {
            background-color: #ffffff; /* White for a clean look against the grey background */
            margin: -20px;
            margin-bottom: 20px;
            padding: 20px;
            border-radius: 0 0 10px 10px; /* Rounded bottom edges */
            color: #4a4a4a; /* Soft grey text for a subtle contrast */
            box-shadow: 0 2px 5px rgba(0, 0, 0, 0.15); /* Enhanced shadow for better depth */
            font-size: 19px; /* Increased font size for better readability */
        }

    </style>
</head>
<body>
    <div class="header">
        <h1>Data Resources for Proteogenomics-related Studies</h1>
    </div>
    <div class="intro">
    <p>This collection of online resources offers brief overviews and easy access to leading databases in the fields of proteomics and proteogenomics research. It covers various research repositories, from raw mass spectrometry data files to preprocessed multi-omics abundance tables.</p>
    </div>

    <div class="container">
        <div class="card">
            <h2>1. CPTAC Pan-Cancer Data</h2>
            <p>This page provides information about the data generated by the National Cancer Institute’s Clinical Proteomic Tumor Analysis Consortium (CPTAC) through the application of large-scale proteomic and genomic analysis to conduct a comprehensive and interconnected proteogenomic characterization of the most prevalent 10 types of cancer. The Supplementary Information section of the page provides download link for 11 different data types (global proteomics, 4 different types of post-translation-modification proteomics, RNAseq, miRNA-seq, CNV, mutation, Methylation Array, clinical and other metadata) for >1000 tumors. All omics data files are preprocessed and carefully QCed abundance tables without access constrain.</p>
            <a href="https://proteomic.datacommons.cancer.gov/pdc/cptac-pancancer">CPTAC Pan-Cancer Data</a>
        </div>

        <div class="card">
            <h2>2. Cancer Cell Lin Encyclopedia Proteomics Data</h2>
            <p>This website is for a comprehensive proteomics study for 375 cell lines from diverse lineages. The Normalized and Other Data section provides links for data table files for peptide-level and protein-level quantification, as well as intermediate analysis results (e.g. correlation between protein abundances with mutations and RNA expressions).</p>
            <a href="https://gygi.hms.harvard.edu/publications/ccle.html">Cancer Cell Lin Encyclopedia Proteomics Data</a>
        </div>

        <div class="card">
            <h2>3. ProteomeXchange</h2>
            <p>ProteomeXchange is a global consortium designed to facilitate the public sharing and dissemination of proteomics data. This website provides a unified framework for researchers to submit, access, and explore high-quality proteomics datasets, ensuring transparency and reproducibility in scientific research. ProteomeXchange supports standardized data submission and retrieval, enabling seamless integration with downstream analysis tools. </p>
            <a href="https://www.proteomexchange.org/">ProteomeXchange</a>
        </div>

        <div class="card">
            <h2>4. MassIVE</h2>
            <p>MassIVE is a community resource developed by the NIH-funded Center for Computational Mass Spectrometry to promote the global, free exchange of mass spectrometry data. MassIVE datasets can be assigned ProteomeXchange accessions to satisfy publication requirements.</p>
            <a href="https://massive.ucsd.edu/ProteoSAFe/static/massive.jsp">MassIVE</a>
        </div>

        <div class="card">
            <h2>5. PeptideAtlas</h2>
            <p>PeptideAtlas is a comprehensive resource for the global proteomics community, offering a curated repository of high-quality mass spectrometry-based proteomics data. It aggregates and standardizes peptide and protein identifications from a wide range of experiments, creating an accessible and reliable reference database. By mapping peptides to annotated genomes and proteomes, PeptideAtlas supports researchers in understanding protein expression, modification, and function across diverse biological contexts.</p>
            <a href="https://peptideatlas.org/">PeptideAtlas</a>
        </div>

        <div class="card">
            <h2>6. PRIDE (PRoteomics IDEntification database)</h2>
            <p>PRIDE (PRoteomics IDEntifications Database) is a leading public repository for proteomics data, maintained by the European Bioinformatics Institute (EMBL-EBI). It serves as a central hub for the deposition, sharing, and exploration of mass spectrometry-based proteomics datasets. PRIDE supports open science by enabling researchers to store and access protein and peptide identifications, post-translational modifications, and quantitative data.</p>
            <a href="https://www.ebi.ac.uk/pride/">PRIDE</a>
        </div>

        <div class="card">
            <h2>7. NCI Proteomic Data Commons</h2>
            <p>The objectives of the National Cancer Institute’s Proteomic Data Commons (PDC) are (1) to make cancer-related proteomic datasets easily accessible to the public, and (2) facilitate direct multiomics integration in support of precision medicine through interoperability with accompanying data resources (genomic and medical image datasets). For each study hosted on this page, users can find a variety of data type files including both proteomics and genomics. For the proteomics data, in most cases, it hosts both raw and processed data. </p>
            <a href="https://proteomic.datacommons.cancer.gov/pdc/">NCI Proteomic Data Commons</a>
        </div>

        <div class="card">
            <h2>8. Proteome-Phenome Atlas </h2>
            <p>This website is a public repository of protein-phenotype associations, providing full results of the phenotypic and genomic associations between ~3,000 plasma proteins and ~1,000 health-related phenotypes for ~50,000 adults in the UK Biobank.</p>
            <a href="https://proteome-phenome-atlas.com/">Proteome-Phenome Atlas</a>
        </div>

    </div>
</body>
</html>
